src/vm-vcpu-ref/src/x86_64/mptable.rs

```diff
@@ -112,229 +117,266 @@ fn mpf_intel_compute_checksum(v: &mpspec::mpf_intel) -> u8 {
+        if cpu_num > MAX_SUPPORTED_CPUS {
+            return Err(Error::TooManyCpus);
+        }
 
-    let mp_size = compute_mp_size(num_cpus);
+        Ok(MpTable {
+            irq_num: IRQ_MAX + 1,
+            cpu_num,
+        })
+    }
 
-    let mut checksum: u8 = 0;
-    let ioapicid: u8 = num_cpus + 1;
+    // Returns the size of this MP table based on its configuration.
+    fn size(&self) -> usize {
+        mem::size_of::<MpfIntel>()
+            + mem::size_of::<MpcTable>()
+            + mem::size_of::<MpcCpu>() * (self.cpu_num as usize)
+            + mem::size_of::<MpcIoapic>()
+            + mem::size_of::<MpcBus>()
+            + mem::size_of::<MpcIntsrc>() * (self.irq_num as usize)
+            + mem::size_of::<MpcLintsrc>() * 2
+    }
 
-    // The checked_add here ensures the all of the following base_mp.unchecked_add's will be without
-    // overflow.
-    if let Some(end_mp) = base_mp.checked_add((mp_size - 1) as u64) {
-        if !mem.address_in_range(end_mp) {
-            return Err(Error::NotEnoughMemory);
+    /// Write the MP Table in guest memory.
+    pub fn write<M: GuestMemory>(&self, mem: &M) -> Result<()> {
+        // Used to keep track of the next base pointer into the MP table.
+        let mut base_mp = GuestAddress(MPTABLE_START);
+        let mp_size = self.size();
+        // The checked_add here ensures the all of the following base_mp.unchecked_add's will be without
+        // overflow.
+        if let Some(end_mp) = base_mp.checked_add((mp_size - 1) as u64) {
+            if !mem.address_in_range(end_mp) {
+                return Err(Error::NotEnoughMemory);
+            }
+        } else {
+            return Err(Error::AddressOverflow);
         }
-    } else {
-        return Err(Error::AddressOverflow);
-    }
 
-    mem.read_from(base_mp, &mut io::repeat(0), mp_size)
-        .map_err(|_| Error::Clear)?;
-
-    {
-        let mut mpf_intel = MpfIntelWrapper(mpspec::mpf_intel::default());
-        let size = mem::size_of::<MpfIntelWrapper>() as u64;
-        mpf_intel.0.signature = SMP_MAGIC_IDENT;
-        mpf_intel.0.length = 1;
-        mpf_intel.0.specification = 4;
-        mpf_intel.0.physptr = (base_mp.raw_value() + size) as u32;
-        mpf_intel.0.checksum = mpf_intel_compute_checksum(&mpf_intel.0);
-        mem.write_obj(mpf_intel, base_mp)
-            .map_err(|_| Error::WriteMpfIntel)?;
-        base_mp = base_mp.unchecked_add(size);
-    }
+        let mut checksum: u8 = 0;
+        let max_ioapic_id: u8 = self.cpu_num + 1;
+
+        mem.read_from(base_mp, &mut io::repeat(0), mp_size)
+            .map_err(|_| Error::Clear)?;
+
+        {
+            let size = mem::size_of::<MpfIntel>() as u64;
+            let mut mpf_intel = MpfIntel(mpspec::mpf_intel {
+                signature: SMP_MAGIC_IDENT,
+                length: 1,
+                specification: MPC_SPEC as u8,
+                physptr: (base_mp.raw_value() + size) as u32,
```

Comment:
Can we add a comment here explaining why the conversion from u64 to u32 is safe?